### DISCUSSION TITLE
지역 기반 실시간 채팅 서비스에서 특정 지역에 갑작스러운 비로 인한 Redis 기반 트래픽 급증 대응 및 최적화 전략

### DISCUSSION THREAD

**Discussion by bysoyeon** [100-hours-a-week/discussions#54] [Deepdive]
### 현상:
지역 기반 실시간 채팅 서비스에서는 사용자 활동이 특정 지역의 상황에 따라 급증하는 경우가 있습니다. 예를 들어, 특정 지역에 갑작스럽게 비가 오는 상황에서는 해당 지역의 사용자들이 실시간으로 우산 판매, 비와 관련된 정보 공유, 지역 커뮤니티 활동 등을 위해 채팅방에 몰리게 됩니다. 이러한 트래픽 급증은 서버 부하를 초래하고, 실시간 데이터 처리에 어려움을 줄 수 있습니다.

### 문제점:
1. **트래픽 급증으로 인한 서버 과부하**:
   - 특정 지역에 비가 오면 그 지역의 사용자들이 한꺼번에 채팅방에 몰리면서 트래픽이 폭발적으로 증가합니다. 이로 인해 서버에 과부하가 발생하고, 응답 지연이나 서비스 중단 등의 문제가 생길 수 있습니다.

2. **Redis 메모리 관리 문제**:
   - Redis를 채팅 메시지 저장소로 사용할 때, 갑작스럽게 메시지가 몰리면 메모리 사용량이 급증하여 메모리 부족 문제가 발생할 수 있습니다. 특히 TTL(Time-To-Live) 설정이 적절하지 않거나 캐시 전략이 비효율적일 경우, 오래된 데이터가 제대로 정리되지 않아 Redis의 성능이 저하될 수 있습니다.

3. **데이터 일관성 문제**:
   - 실시간 데이터 처리에서 중요한 점은 일관된 데이터를 사용자에게 제공하는 것입니다. 하지만 트래픽이 급증하면서 메시지 순서가 꼬이거나, 데이터가 제대로 반영되지 않는 문제가 발생할 수 있습니다. 이는 특히 Redis를 클러스터링 환경에서 사용할 때 발생할 수 있으며, 데이터 동기화의 어려움을 초래합니다.

4. **서비스 지연과 사용자 경험 저하**:
   - 사용자가 몰리면서 응답 시간이 길어지거나, 메시지 전송이 지연되면 실시간 채팅 서비스의 사용자 경험이 크게 저하됩니다. 실시간성이 중요한 채팅 서비스에서 지연은 사용자의 이탈을 초래할 수 있습니다.

5. **지역별 트래픽 집중으로 인한 부하 불균형**:
   - 채팅 서비스가 지역 기반으로 운영되기 때문에, 특정 지역에서 트래픽이 몰릴 경우 해당 지역의 서버나 데이터베이스가 과부하 상태가 됩니다. 반면, 다른 지역의 리소스는 비활성화되어 비효율적인 자원 사용이 발생할 수 있습니다.

**Comment by ej31** [comment]
스읍.... 레디스 터트리기.... 쉽지 않은데... 🤔
이 주제 다루실거라면 꼭 모니터링 툴과 redis export, grafana 등의 툴을 활용해서 그래프 쫙쫙 깔아주셔야 합니다.
